src/cloudevent/mocks/firestore/helpers.ts

```diff
@@ -0,0 +1,147 @@
+    database,
+    namespace,
+    document,
+    params,
+
+    data
+  }
+}
+
+export function getDocumentSnapshotChangeCloudEvent(
+  cloudFunction: CloudFunction<firestore.FirestoreEvent<Change<DocumentSnapshot>>>,
+  cloudEventPartial?: DeepPartial<firestore.FirestoreEvent<Change<DocumentSnapshot> | ChangeLike>>,
+) {
+  const {
+    location,
+    project,
+    database,
+    namespace,
+    document,
+    params
+  } = getFirestoreEventFields(cloudFunction, cloudEventPartial);
+  const data = getOrCreateDocumentSnapshotChange(cloudEventPartial?.data, document);
+  return {
+    ...getBaseCloudEvent(cloudFunction),
+
+    location,
+    project,
+    database,
+    namespace,
+    document,
+    params,
+
+    data
+  }
+}
+
+function getFirestoreEventFields(
+  cloudFunction: CloudFunction<
+    firestore.FirestoreEvent<
+      DocumentSnapshot | Change<DocumentSnapshot>
+    >
+  >,
+  cloudEventPartial?: DeepPartial<
+    firestore.FirestoreEvent<
+      DocumentSnapshot | Change<DocumentSnapshot>
+    >
+  >,
+) {
+  const location = cloudEventPartial?.location || 'us-central1';
+
+  const project = 
+    cloudEventPartial?.project || 
+    process.env.GCLOUD_PROJECT ||
+    'testproject';
+
+  const databaseOrExpression = 
+    cloudEventPartial?.database || 
+    cloudFunction?.__endpoint?.eventTrigger?.eventFilters?.database ||
+    '(default)';
+  const database = resolveStringExpression(databaseOrExpression);
+
+  const namespaceOrExpression = 
+    cloudEventPartial?.namespace || 
+    cloudFunction?.__endpoint?.eventTrigger?.eventFilters?.namespace ||
+    '(default)';
+  const namespace = resolveStringExpression(namespaceOrExpression);
+
+  const params = cloudEventPartial?.params || {};
+
+  const documentOrExpression = 
+    cloudEventPartial?.document || 
+    cloudFunction?.__endpoint?.eventTrigger?.eventFilters?.document ||
+    cloudFunction?.__endpoint?.eventTrigger?.eventFilterPathPatterns?.document || 
+    '/foo/bar';  
+  const documentRaw = resolveStringExpression(documentOrExpression);
+  const document = extractRef(documentRaw, params);
+
+  return {
+    location,
+    project,
+    database,
+    namespace,
+    document,
+    params
+  };
+}
+
+function getOrCreateDocumentSnapshot(
+  data: DocumentSnapshot | object,
+  ref: string,
+) {
+  if (data instanceof DocumentSnapshot) {
+    return data;
+  }
+  if (data instanceof Object) {
+    return makeDocumentSnapshot(data, ref);
+  }
+  return exampleDocumentSnapshot();
+}
+
+function getOrCreateDocumentSnapshotChange(
+  data: DeepPartial<Change<DocumentSnapshot> | ChangeLike>,
+  ref: string,
+) {
+  if (data instanceof Change) {
+    return data;
+  }
+  // If only the "before" or "after" is specified (to simulate a
+  // Created or Deleted event for the onWritten trigger), then we 
+  // include the user's before/after object in the mock event and
+  // use an example snapshot for the other.
+  if (data instanceof Object && (data.before || data.after)) {
```

Comment:
is this || suppose to be a && ? Does `getOrCreateDocumentSnapship` still work?

Reply:
I used || instead of && because I thought it may be surprising for users who specify an object like

```
const partial = {
  data: {
    before: {
      foo: 'bar',
    },
}
```
and expect the generated mock event data to include `event.data.before`, but if we use &&, they will be required to specify both the `before` and `after` objects, and if they don't, the data they provide will be overwritten and they will see the example data in the cloud event instead.